File: WPF Databinding in Visual Studio 2010 RTM/[C#]-WPF Databinding in Visual Studio 2010 RTM/C#/WpfEfSampleApp/OrdersCollection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using WpfEfDAL;
using System.Collections.ObjectModel;
using System.ComponentModel;

namespace WpfEfSampleApp
{
    class OrdersCollection : MyBaseCollection<Order>
    {
        /// <summary>
        // Return true if any order in this collection has a validation error
        /// <summary>
        public Boolean HasErrors
        {
            get {
                return ( (from order in this where order.HasErrors select order).Count() > 0 );
            }
        }

        public OrdersCollection (IEnumerable<Order> orders, OMSEntities context) : base(orders, context)
        {
        }

        protected override void InsertItem(int index, Order item)
        {
            item.OrderDetails.AssociationChanged += new CollectionChangeEventHandler(Details_CollectionChanged);

            if (! this.IsLoading)
                this.Context.AddToOrders(item);

            base.InsertItem(index, item);
        }

        protected override void RemoveItem(int index)
        {
            Order order = this[index];

            order.OrderDetails.AssociationChanged -= new CollectionChangeEventHandler(Details_CollectionChanged);

            for (int i = 0; i <= order.OrderDetails.Count; i++)
            {
                // When deleting an order, delete any details if any exist
                this.Context.DeleteObject(order.OrderDetails.ElementAt(0));
            }

            this.Context.DeleteObject(order);
            base.RemoveItem(index);
        }

        private void Details_CollectionChanged(object sender, CollectionChangeEventArgs e)
        {
            if (e.Action == CollectionChangeAction.Remove)
            {
                // Adding a detail to an order is handled automatically for us but we need to tell the ObjectContext 
                // to delete the detail if a detail is removed from the OrderDetails EntityCollection 
                this.Context.DeleteObject((OrderDetail)e.Element);
            }
        }
    }
}

File: WPF Databinding in Visual Studio 2010 RTM/[VB]-WPF Databinding in Visual Studio 2010 RTM/VB/WpfEfSampleApp/OrdersCollection.vb
Imports WpfEfDAL
Imports System.Collections.ObjectModel
Imports System.ComponentModel

Public Class OrdersCollection
    Inherits MyBaseCollection(Of Order)

    ''' <summary>
    ''' Return true if any order in this collection has a validation error
    ''' </summary>
    Public ReadOnly Property HasErrors() As Boolean
        Get
            Return (Aggregate order In Me Where order.HasErrors Into Count()) > 0
        End Get
    End Property

    Sub New(ByVal orders As IEnumerable(Of Order), ByVal context As OMSEntities)
        MyBase.New(orders, context)
    End Sub

    Protected Overrides Sub InsertItem(ByVal index As Integer, ByVal item As Order)
        AddHandler item.OrderDetails.AssociationChanged, AddressOf Details_CollectionChanged
        If Not Me.IsLoading Then
            Me.Context.AddToOrders(item)
        End If
        MyBase.InsertItem(index, item)
    End Sub

    Protected Overrides Sub RemoveItem(ByVal index As Integer)
        Dim order = Me(index)
        RemoveHandler order.OrderDetails.AssociationChanged, AddressOf Details_CollectionChanged
        For i = order.OrderDetails.Count - 1 To 0 Step -1
            'When deleting an order, delete any details if any exist
            Me.Context.DeleteObject(order.OrderDetails(i))
        Next
        Me.Context.DeleteObject(order)
        MyBase.RemoveItem(index)
    End Sub

    Private Sub Details_CollectionChanged(ByVal sender As Object, ByVal e As CollectionChangeEventArgs)
        If e.Action = CollectionChangeAction.Remove Then
            'Adding a detail to an order is handled automatically for us but we need to tell the ObjectContext 
            ' to delete the detail if a detail is removed from the OrderDetails EntityCollection 
            Me.Context.DeleteObject(CType(e.Element, OrderDetail))
        End If
    End Sub

End Class
